app/services/course/assessment/question/programming/cpp/cpp_package_service.rb

```diff
@@ -0,0 +1,254 @@
+      zip.print "\n"
+
+      zip.put_next_entry 'tests/prepend.cc'
+      zip.print "\n"
+      zip.print File.read(AUTOGRADE_INCLUDE_PATH)
+      zip.print "\n"
+      zip.print @test_params[:prepend]
+      zip.print "\n"
+      zip.print File.read(AUTOGRADE_PRE_PATH)
+      zip.print "\n"
+
+      zip.put_next_entry 'tests/autograde.cc'
+
+      [:public, :private, :evaluation].each do |test_type|
+        zip_test_files(test_type, zip)
+      end
+
+      zip.print "\n"
+      zip.print File.read(AUTOGRADE_POST_PATH)
+
+      # Creates Makefile
+      zip.put_next_entry 'Makefile'
+      zip.print File.read(MAKEFILE_PATH)
+
+      zip.put_next_entry '.meta'
+      zip.print @meta.to_json
+    end
+
+    Zip::File.open(tmp.path) do |zip|
+      @test_params[:data_files].try(:each) do |file|
+        next if file.nil?
+        zip.add(file.original_filename, file.tempfile.path)
+      end
+
+      data_files_to_keep.each do |file|
+        zip.add(File.basename(file.path), file.path)
+      end
+    end
+
+    tmp
+  end
+
+  def zip_test_files(test_type, zip) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+    tests = @test_params[:test_cases]
+    tests[test_type]&.each&.with_index(1) do |test, index|
+      # String types should be displayed with quotes, other types will be converted to string
+      # with the str method.
+      expected = string?(test[:expected]) ? test[:expected].inspect : "#{test[:expected]}"
+      hint = test[:hint].blank? ? String(nil) : "RecordProperty(\"hint\", #{test[:hint].inspect})"
+
+      test_fn = <<-CPlusPlus
+        TEST(Autograder, test_#{test_type}_#{format('%02i', index)}) {
+          RecordProperty("expression", #{test[:expression].inspect});
+          custom_evaluation(#{test[:expression]}, #{test[:expected]});
+          #{hint};
+        }
+      CPlusPlus
+
+      zip.print test_fn
+    end
+  end
+
+  def get_data_files_meta(data_files_to_keep, new_data_files)
+    data_files = []
+
+    new_data_files.each do |file|
+      sha256 = Digest::SHA256.file(file.tempfile).hexdigest
+      data_files.append(filename: file.original_filename, size: file.tempfile.size, hash: sha256)
+    end
+
+    data_files_to_keep.each do |file|
+      sha256 = Digest::SHA256.file(file).hexdigest
+      data_files.append(filename: File.basename(file.path), size: file.size, hash: sha256)
+    end
+
+    data_files.sort_by { |file| file[:filename].downcase }
+  end
+
+  def generate_meta(data_files_to_keep)
+    meta = default_meta
+
+    [:submission, :solution, :prepend, :append].each { |field| meta[field] = @test_params[field] }
+
+    new_data_files = (@test_params[:data_files] || []).reject(&:nil?)
+    meta[:data_files] = get_data_files_meta(data_files_to_keep, new_data_files)
+
+    [:public, :private, :evaluation].each do |test_type|
+      meta[:test_cases][test_type] = @test_params[:test_cases][test_type] || []
+    end
+
+    meta.as_json
+  end
+
+  def test_params(params)
```

Comment:
I think you can shift this to the superclass too. The subclasses have the ability to overwrite this if they need a slightly different param.